10up/ElasticPress
# Decouple Instant Results from the Modal

### Description of the Change
This change decouples Instant Results from the modal UI. This is achieved by moving the state logic out of the modal to an `ApiSearchProvider` provider component and providing a custom hook `useApiSearch()` for consuming its state and methods. The various Instant Results components have been updated to use this hook instead of the generic Context.

The provider component is not currently publicly exposed, but that's something that could be possible in a future version once its API is more stable.

The main challenge was how to handle the open/closed state of the modal,  since this was coupled with the history API functionality so that back and forward could properly open and close the modal. The problem being that this open/close logic is related to the modal UI and not the rest of the state. The solution I came up with was to have a generic on/off state of the provider that the Modal UI could control and whose state it could reflect to open and close the modal, while other UI implementations could theoretically use it differently.

A side effect of some of the improvements made here is that a bug where an Instant Results request would be sent on every page load, even if the modal was never opened, has been fixed. Now an API request is only sent when the modal is opened.


Closes #2979

### How to test the Change
1. Confirm that when Instant Results is enabled and a page is loaded, no AJAX request is sent to the API until the modal is opened.
2. There are no other user-facing changes, so the only testing required is to confirm that everything in Instant Results continues to function as expected.

### Changelog Entry
Changed - Under the hood improvements to the structure of Instant Results.
Fixed - An issue where API requests for Instant Results would be sent on page load before the modal has been opened.

### Credits
@JakePT

### Checklist:
